File: hyeyoung/week5/햄버거만들기.java
package hyeyoung.week5;

import java.util.LinkedList;
import java.util.Stack;

public class 햄버거만들기 {
}

class Solution {
    public int solution(int[] ingredient) {
        int answer = 0;
        Stack<Integer> stack = new Stack<>();
        for (int in : ingredient) {
            stack.push(in);
            if (stack.size() > 3) {
                int size = stack.size();
                if(stack.get(size - 1) == 1
                        && stack.get(size - 2) == 3
                        && stack.get(size - 3) == 2
                        && stack.get(size - 4) == 1) {
                    answer++;
                    stack.pop();
                    stack.pop();
                    stack.pop();
                    stack.pop();
                }
            }
        }
        return answer;
    }
}

File: SiHyeon/week5/햄버거 만들기.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> ingredient) {
    int answer = 0;
    vector<int> buger = { 1,2,3,1 };
    int index = 0;
    for (int i = 0; i < ingredient.size(); i++)
    {
        if (ingredient[i] == 1 && i + buger.size() <= ingredient.size())
            for (int j = 0; j < buger.size(); j++)
            {
                if (buger[j] != ingredient[i + j]) //시작재료가 아니거나 만들 재료가 부족하거나 재료가 다르면 넘어감
                    break;
                else if (j == buger.size() - 1) //끝까지 했으면 완성
                {
                    ingredient.erase(ingredient.begin() + i, ingredient.begin() + i + j + 1);
                    i = i - 3 > 0 ? i - 3 : -1; //완성을 하면 맨처음부터 다시 찾는것이 아니라 완성된 위치에서 뒤로 2번째부터 찾으면 된다. 그보다 전것은 이미 못만드는 것이기 때문
                    answer++;
                }
            }
    }
    return answer;
}